src/components/ViewerMessaging/MessagingArgument.tsx

```diff
@@ -0,0 +1,105 @@
+    name: string,
+    schema: MessageSchema
+): Definition | undefined {
+    const trimmedName = trimDefinitionsName(name);
+    return schema.definitions[trimmedName];
+}
+
+export default function MessagingArgument(props: MessagingArgumentProps) {
+    const { schema } = props;
+
+    let definition: Definition | string | undefined = props.definition;
+
+    if (typeof definition === "string") {
+        const foundDefinition = getReferencedDefinition(name, schema);
+        console.warn("Couldn't find definition:", definition);
+        definition = foundDefinition;
+    }
+
+    if (!definition) {
+        return <code>null</code>;
+    }
+    // This is a single type referencing another definition
+    else if (definition.$ref) {
+        const referencedDef = getReferencedDefinition(definition.$ref, schema);
+
+        // We only hyperlink to object type definitions, everything else can be inlined.
+        if (referencedDef && referencedDef.type === "object") {
+            return <MessagingRef name={definition.$ref} schema={schema} />;
+        } else if (referencedDef) {
+            return (
+                <MessagingArgument definition={referencedDef} schema={schema} />
+            );
+        }
+
+        return <MessagingRef name={definition.$ref} schema={schema} />;
+    }
+    // This is a single type
+    else if (definition.type) {
+        if (definition.type === "string" && definition.enum) {
+            const enumType = (definition.enum as PrimitiveType[])
+                .map((val) => `"${val}"`)
+                .join(" | ");
+            return <code>{enumType}</code>;
+        } else if (definition.type === "array" && definition.items) {
+            if (Array.isArray(definition.items)) {
+                return (
+                    <>
+                        <div>Any of:</div>
+                        {definition.items.map((option, index) => (
+                            // There's not a guaranteed safe identifier we can use for the key prop, fall back to index.
+                            <div key={option.$ref || index}>
+                                <MessagingArgument
+                                    definition={option}
+                                    schema={schema}
+                                />
+                                []
+                            </div>
+                        ))}
+                    </>
+                );
+            }
+            if (definition.items && (definition.items as Definition).$ref) {
+                const itemsRef = (definition.items as Definition).$ref!;
+                return <MessagingRef isArray name={itemsRef} schema={schema} />;
+            }
+            return <code>{(definition.items as Definition).type}[]</code>;
+        } else if (definition.type === "object") {
+            // We don't support rendering object type inline, should only reference by link.
+            return <code>unknown</code>;
```

Comment:
I noticed the objects with "additionalProperties" are getting swallowed by this lack of support in particular... For something like the `results.convert-to-csv` command, this `type="object"` style argument could be quite important, as it contains ways to set properties like the delimiter, or numberFormat, etc.

Could we make a ticket to capture this work (probably makes the most sense to do it after 5.7 changes, or this comment might not even be applicable after 5.7 changes) so it doesn't get lost?

Reply:
Hmm interesting. I think something may be going astray on the schema generation side. I don't think that should actually be there. @mheavenor any thoughts as to what's going on? Here's an example:

```json
{
        "results.convert-to-csv": {
            "description": "",
            "input": {
                "anyOf": [
                    {
                        "$ref": "#/definitions/@geocortex.api.data.FeatureSet.FeatureSet"
                    },
                    {
                        "$ref": "#/definitions/@geocortex.api.data.Feature.Feature"
                    },
                    {
                        "$ref": "#/definitions/@geocortex.api.data.FeatureList.FeatureList"
                    },
                    {
                        "$ref": "#/definitions/@geocortex.api.data.FeatureStream.FeatureStream"
                    },
                    {
                        "$ref": "#/definitions/Iterable"
                    },
                    {
                        "$ref": "#/definitions/esri.Collection"
                    },
                    {
                        "$ref": "#/definitions/Results"
                    },
                    {
                        "additionalProperties": false,
                        "properties": {
                            "alwaysQuote": {
                                "description": "Whether to always quote values, even if they do not require it. The default is false.",
                                "type": "boolean"
                            },
                            "dateFormat": {
                                "description": "The format for dates. The default is DateFormat.DEFAULT.",
                                "type": "string"
                            },
                            "delimiter": {
                                "description": "The delimiter between fields. The default is \",\".",
                                "type": "string"
                            },
                            "escapeFormulaChars": {
                                "description": "Whether to a prefix values that start with typical formula triggering chars (=,-,+,@).",
                                "type": "boolean"
                            },
                            "features": {
                                "$ref": "#/definitions/FeaturesLike",
                                "description": "The feature(s) to convert."
                            },
                            "geometryFormat": {
                                "$ref": "#/definitions/@geocortex.api.data.convert.GeometryFormat",
                                "description": "The format for serializing the geometry. The default is WKT."
                            },
                            "inSpatialReference": {
                                "$ref": "#/definitions/esri.SpatialReference",
                                "description": "When importing from CSV, this is the spatial reference of the geometry in the CSV."
                            },
                            "includeByteOrderMark": {
                                "description": "Whether to include a Unicode byte order mark at the beginning of the text. True by default.",
                                "type": "boolean"
                            },
                            "includeHeaderRow": {
                                "description": "Whether to include a header row with field names. Default is true.",
                                "type": "boolean"
                            },
                            "numberFormat": {
                                "description": "The format for numbers. The default is NumberFormat.DEFAULT.",
                                "type": "string"
                            },
                            "outFields": {
                                "description": "A list of fields to include. The default is \"*\", meaning all fields.",
                                "items": {
                                    "type": "string"
                                },
                                "type": "array"
                            },
                            "outSpatialReference": {
                                "$ref": "#/definitions/esri.SpatialReference",
                                "description": "When exporting to CSV, this is the spatial reference of the geometry in the CSV. The default is WGS84. When importing from CSV, this is the spatial reference of the geometry in the result."
                            },
                            "rowDelimiter": {
                                "description": "The delimiter between rows. The default is the platform-specific line ending (e.g. \"\\r\\n\" on Windows, \"\\n\" on Unix or MacOS).",
                                "type": "string"
                            }
                        },
                        "required": [
                            "features"
                        ],
                        "type": "object"
                    }
                ]
            },
            "output": {
                "additionalProperties": false,
                "properties": {
                    "name": {
                        "description": "A Filename representation for the conversion.",
                        "type": "string"
                    },
                    "result": {
                        "$ref": "#/definitions/Blob",
                        "description": "The result of the conversion in the specified form."
                    }
                },
                "required": [
                    "name",
                    "result"
                ],
                "type": "object"
            }
        }
}
```

Reply:
Another strange one in `GeocodeArgs`:

```json
{
 "GeocodeArgs": {
            "additionalProperties": false,
            "properties": {
                "address": {
                    "anyOf": [
                        {
                            "$ref": "#/definitions/@geocortex.api.data.Feature.Feature"
                        },
                        {
                            "additionalProperties": {
                            },
                            "type": "object"
                        },
                        {
                            "items": {
                                "anyOf": [
                                    {
                                        "$ref": "#/definitions/@geocortex.api.data.Feature.Feature"
                                    },
                                    {
                                        "additionalProperties": {
                                        },
                                        "type": "object"
                                    },
                                    {
                                        "type": "string"
                                    }
                                ]
                            },
                            "type": "array"
                        },
                        {
                            "type": "string"
                        }
                    ]
                },
                "options": {
                    "$ref": "#/definitions/@geocortex.api.tasks.geocoding.Geocoder.GeocodeOptions"
                }
            },
            "required": [
                "address",
                "options"
            ],
            "type": "object"
        }
}
```

Reply:
@Noah-Silvera - Michael will be resolving most of this in 5.7. There's an outlier for The `GeocodeArgs` example i posted, but there's not a great solution at this time, so we're okay leaving it as is. The others should be fixed up though.

Reply:
Sounds good to me!